
![titanic.jpg](https://image.geo.de/31780898/t/gt/v3/w960/r1.5/-/titanic---pa--122855442.jpg)
# Midterm Test 

Hello!

You already learned a lot how to explore and understand data. Let's test this! If you hand in your solution we will check it so you get direct feedback and know where you stand. To get feedback save a copy and rename it, so it contains your name. Then save the copy to this 
[folder](https://drive.google.com/drive/folders/1JqOTlRt9VhK3NVHt0svBHLI8DnMdpp5Q?usp=sharing).

The deadline for handing in your solution is May 9.

You can answer those questions as you wish by just coding or inspecting the data through a plot. But please answer the question with a sentence or two where you explain what the result is telling you. 

If you have any questions, please ask us on slack. We are happy to help.

Have fun coding!

### Titanic Dataset

Today we will explore a new dataset, the infamous Titanic dataset. It contains informations on passengers of the Titanic. At kaggle many people already investigated this dataset. Check it out to get new ideas: -> https://www.kaggle.com/c/titanic/overview


The dataset contains 12 columns.

- survival: 	Survival 	0 = No, 1 = Yes
- pclass: 	Ticket class 	1 = 1st, 2 = 2nd, 3 = 3rd
- sex: 	Sex 	
- Age: 	Age in years 	
- sibsp: 	# of siblings / spouses aboard the Titanic 	
- parch: 	# of parents / children aboard the Titanic 	
- ticket: 	Ticket number 	
- fare: 	Passenger fare 	
- cabin: 	Cabin number 	
- embarked: 	Port of Embarkation 	C = Cherbourg, Q = Queenstown, S = Southampton

It can be downloaded from: https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv


### 1. Load Libraries and data

Load libraries you might need and load the data using pandas. Get a first overview:
- Is the data loaded correctly?
- How does it look like?
- How many rows, how many columns?
- Do we see any Null values? What could they mean?

In [1]:
import pandas as pd
import seaborn as sns
import numpy as np

In [2]:
df = pd.read_csv("https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv")

In [3]:
df.head(50)



PassengerId  Survived  Pclass  \
0             1         0       3   
1             2         1       1   
2             3         1       3   
3             4         1       1   
4             5         0       3   
5             6         0       3   
6             7         0       1   
7             8         0       3   
8             9         1       3   
9            10         1       2   
10           11         1       3   
11           12         1       1   
12           13         0       3   
13           14         0       3   
14           15         0       3   
15           16         1       2   
16           17         0       3   
17           18         1       2   
18           19         0       3   
19           20         1       3   
20           21         0       2   
21           22         1       2   
22           23         1       3   
23           24         1       1   
24           25         0       3   
25           26         1       3   
26           27         0       3   
27           28         0       1   
28           29         1       3   
29           30         0       3   
30           31         0       1   
31           32         1       1   
32           33         1       3   
33           34         0       2   
34           35         0       1   
35           36         0       1   
36           37         1       3   
37           38         0       3   
38           39         0       3   
39           40         1       3   
40           41         0       3   
41           42         0       2   
42           43         0       3   
43           44         1       2   
44           45         1       3   
45           46         0       3   
46           47         0       3   
47           48         1       3   
48           49         0       3   
49           50         0       3   

                                                 Name     Sex   Age  SibSp  \
0                             Braund, Mr. Owen Harris    male  22.0      1   
1   Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                              Heikkinen, Miss. Laina  female  26.0      0   
3        Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                            Allen, Mr. William Henry    male  35.0      0   
5                                    Moran, Mr. James    male   NaN      0   
6                             McCarthy, Mr. Timothy J    male  54.0      0   
7                      Palsson, Master. Gosta Leonard    male   2.0      3   
8   Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                 Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   
10                    Sandstrom, Miss. Marguerite Rut  female   4.0      1   
11                           Bonnell, Miss. Elizabeth  female  58.0      0   
12                     Saundercock, Mr. William Henry    male  20.0      0   
13                        Andersson, Mr. Anders Johan    male  39.0      1   
14               Vestrom, Miss. Hulda Amanda Adolfina  female  14.0      0   
15                   Hewlett, Mrs. (Mary D Kingcome)   female  55.0      0   
16                               Rice, Master. Eugene    male   2.0      4   
17                       Williams, Mr. Charles Eugene    male   NaN      0   
18  Vander Planke, Mrs. Julius (Emelia Maria Vande...  female  31.0      1   
19                            Masselmani, Mrs. Fatima  female   NaN      0   
20                               Fynney, Mr. Joseph J    male  35.0      0   
21                              Beesley, Mr. Lawrence    male  34.0      0   
22                        McGowan, Miss. Anna "Annie"  female  15.0      0   
23                       Sloper, Mr. William Thompson    male  28.0      0   
24                      Palsson, Miss. Torborg Danira  female   8.0      3   
25  Asplund, Mrs. Carl Oscar (Selma Augusta Emilia...  female  38.0      1   
26        

In [4]:
df.shape
#the dataset has 891 rows and 12 columns.

(891, 12)

In [5]:
df.isna().sum()
#there is null values in age columns,cabin, and embarked as well.

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

### 2. What is the average fare per class? On which range do the fares per 

1.   List item
2.   List item

class spread?

In [6]:
df_class_2=df[df['Pclass']==2]
df_class_2.head(10)
print(df_class_2['Fare'].mean())
print(df_class_2['Fare'].min())
print(df_class_2['Fare'].max())

#The average class

20.66218315217391
0.0
73.5


The average far per class 1 is 20.662 and the range is between 0 and **73.5**

In [7]:
df_class_3=df[df['Pclass']==3]
df_class_3.head(10)
print(df_class_3['Fare'].mean())
print(df_class_3['Fare'].min())
print(df_class_3['Fare'].max())

13.675550101832997
0.0
69.55


The average far per class 1 is 13.675 and the range is between 0 and 69.55

In [8]:
df_class_1=df[df['Pclass']==1]
df_class_1.head(10)
print(df_class_1['Fare'].mean())
print(df_class_1['Fare'].min())
print(df_class_1['Fare'].max())

84.15468749999992
0.0
512.3292


The average far per class 1 is 84.154 and the range is between 0 and 512.3292

### 3. How many people had a cabin? 

In [9]:
df['Cabin'].isna().sum()
#the number of null values is 687 so the we can extract the people that had a cabin by calculate the minus between the number of rows in dataset and 687

687

the number of null values is 687 so the we can extract the people that had a cabin by calculate the minus between the number of rows in dataset and 687 
so the number of people that had a cabin are 891 - 687 = 204

4. At which port lived the richest?




In [10]:
df_class_1['Embarked'].value_counts()

S    127
C     85
Q      2
Name: Embarked, dtype: int64

so we can extract from the count  of values in column Embearked that the S value is the highest value the most of the richest people that choosed first calss were living in Southampton 

### 5. What is the most frequent last name?


In [11]:
df['LastName'] = df['Name'].apply(lambda x: x.split(',')[0])
df['LastName'].value_counts()


Andersson    9
Sage         7
Panula       6
Skoog        6
Carter       6
            ..
Hanna        1
Lewy         1
Mineff       1
Haas         1
Dooley       1
Name: LastName, Length: 667, dtype: int64

By splitiing the last name from the column name and after that using the method value_counts we extract that Anderson is the most frequent last name

### 6. Not for every person the age is noted. Fill in the gaps with the average age per gender.

In [21]:
df_male=df[df['Sex']=='male']
df_female=df[df['Sex']=='female']
df_male_age_mean=df_male['Age'].mean()
df_female_age_mean=df_female['Age'].mean()
print(df_male['Age'].mean())
print(df_female['Age'].mean())
    


30.50582424304206
28.216730048707397


In [24]:
df_male['Age'] = df_male['Age'].fillna(df_male_age_mean)
df_female['Age']=df_female['Age'].fillna(df_female_age_mean)
df['Age'].isnull().sum()

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


0

### Bonus: Upload this file as .ipnyb - notebook in your personal git-hub account and share the address here.

In [ ]:
# address: 